cmd/segment-reaper/delete.go

```diff
@@ -53,9 +65,91 @@ func cmdDelete(cmd *cobra.Command, args []string) (err error) {
 		err = errs.Combine(err, inputFile.Close())
 	}()
 
-	_ = csv.NewReader(inputFile)
+	csvReader := csv.NewReader(inputFile)
+	csvReader.FieldsPerRecord = 5
+	csvReader.ReuseRecord = true
+
+	segmentsDeleted := 0
+	segmentsErrored := 0
+	segmentsSkipped := 0
+	for {
+		record, err := csvReader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			log.Error("error while reading record", zap.Error(err))
+			continue
+		}
+
+		projectID := record[0]
+		segmentIndex := record[1]
+		bucketName := record[2]
+		encodedPath := record[3]
+		creationDateFromReport, err := time.Parse(time.RFC3339Nano, record[4])
+		if err != nil {
+			log.Error("error while parsing date", zap.Error(err))
+			continue
+		}
+
+		encryptedPath, err := base64.StdEncoding.DecodeString(encodedPath)
+		if err != nil {
+			log.Error("error while decoding encrypted path", zap.Error(err))
+			continue
+		}
+
+		path := storj.JoinPaths(projectID, segmentIndex, bucketName, string(encryptedPath))
+		rawPath := storj.JoinPaths(projectID, segmentIndex, bucketName, encodedPath)
+
+		err = deleteSegment(ctx, db, path, creationDateFromReport, deleteCfg.DryRun)
+		if err != nil {
+			if errKnown.Has(err) {
+				segmentsSkipped++
+			} else {
+				segmentsErrored++
+			}
+			log.Error("error while deleting segment", zap.String("path", rawPath), zap.Error(err))
+			continue
+		}
+
+		log.Debug("segment deleted", zap.String("path", rawPath))
+		segmentsDeleted++
+	}
+
+	log.Info("summary", zap.Int("deleted", segmentsDeleted), zap.Int("skipped", segmentsSkipped), zap.Int("errored", segmentsSkipped))
+
+	return nil
+}
+
+func deleteSegment(ctx context.Context, db metainfo.PointerDB, path string, creationDate time.Time, dryRun bool) error {
+	pointerBytes, err := db.Get(ctx, []byte(path))
+	if err != nil {
+		return err
+	}
 
-	// TODO logice will be added in next PR
+	pointer := &pb.Pointer{}
+	err = proto.Unmarshal(pointerBytes, pointer)
+	if err != nil {
+		return err
+	}
+
+	// check if pointer has been replaced
+	if !pointer.GetCreationDate().Equal(creationDate) {
+		// pointer has been replaced since detection, do not delete it.
+		return errKnown.New("segment won't be deleted, create date mismatch: %s -> %s", pointer.GetCreationDate(), creationDate)
+	}
+
+	if !dryRun {
+		// delete the pointer using compare-and-swap
+		err = db.CompareAndSwap(ctx, []byte(path), pointerBytes, nil)
+		if storage.ErrValueChanged.Has(err) {
+			// race detected while deleting the pointer, do not try deleting it again.
+			return errKnown.New("segment won't be deleted, race detected while deleting the pointer")
+		}
```

Comment:
We are losing the original error. Can we wrap it instead of replacing it?

Reply:
If we warp it, I would create 2 error classes for distinguishing from this one https://github.com/storj/storj/pull/3660/files#diff-e8fb83f0bab533d6fafa0225792233b5R139 unless that the error returned by `db.CompareAndSwap` already has a message that tells clearly that the error is because of that.

I would love to have something called `errs.Wrapf(err, "here an additional message")`